### PR TITLE
fix(auto-spies-core): don't add getters to method names when creating spies from class

### LINKED ISSUE #40: Spying on observable property getter is failing
**Describe the bug**
While following the example at https://www.npmjs.com/package/jasmine-auto-spies?activeTab=readme#-spying-on-observable-properties, I did created a spy for the `userData$` observable property of the `OidcSecurityService` of `angular-auth-oidc-client`.

I set up the spy like the example and attempted to set up the `nextWith` and got the error `nextWith is not a function`

I have stepped thru the code and at https://github.com/hirezio/auto-spies/blob/master/packages/auto-spies-core/src/create-auto-spy-from-class.ts line 74 before the methodNames are iterated, the spy behaves correctly and it begins to fail after the method names are iterated and it appears to be caused because the observable property, userData$ is contained in the methodNames variable and so it is overwriting the property set up that occurred on line 68: 
` autoSpy[observablePropName] = createObservablePropSpy();`   

**To Reproduce**

1. Create a class that depends on OidcSecurityService from angular-auth-oidc-client
2. Create a jasmine unit test for the class that setups the spy like so:
```
    const oidcSecurityServiceSpy = createSpyFromClass(OidcSecurityService, {
      observablePropsToSpyOn: ['userData$']
    });
    oidcSecurityServiceSpy.userData$.nextWith({
      preferred_username: 'fakeUser'
    });
```

3. Run the test and see the error "oidcSecurityServiceSpy.userData$.nextWith is not a function''

**Expected behavior**
No error when try to assign nextWith result

**Desktop (please complete the following information):**

- win 10
- Browser -chrome 
- Version: 87.0.4280.88 (Official Build) (64-bit)
- jasmine auto spies: tried with 6.6.0 and 6.9.1

**Additional context**
My workaround is to re-assign the prop spy after the call to `createSpyFromClass` by re-using the same code that is used to set it previously on line 68:
`autoSpy[observablePropName] = createObservablePropSpy();`

It seems the problem lies in the population of `methodNames`. Using `Object.getOwnPropertyNames` will retrieve these observable property getters along with all the function properties. Instead you could use `getOwnPropertyDescriptors` and use that to only retrieve the functions or else find the ones that are getters so they can be removed.

### PR DESCRIPTION
**Pull request by GuilleEneas**
fix #40

### CODE REVIEW

File: packages/auto-spies-core/src/create-auto-spy-from-class.ts [hunk 1]
```diff
@@ -85,16 +85,19 @@ function getAllMethodNames(obj: any): string[] {
...
-  /* istanbul ignore else */
-  if (constructorIndex >= 0) {
-    methods.splice(constructorIndex, 1);
-  }
   return methods;
 }
+
+function extractMethodsFromObject(obj: any) {
+  const propertyDescriptors = Object.getOwnPropertyDescriptors(obj);
+  return Object.entries(propertyDescriptors).reduce((names, [name, descriptor]) => {
```
Comment (GuilleEneas) [inline packages/auto-spies-core/src/create-auto-spy-from-class.ts:97]: In terms of browser compatibility I don't know if I should use the old Object.keys and the old for loop

File: packages/jasmine-auto-spies/src/tests/observable-spy-utils.spec.ts [hunk 2]
```diff
@@ -596,6 +596,74 @@ describe('createSpyFromClass - Observables', () => {
       });
     });
 
+    describe(`GIVEN class spy is configured with a manual getter observable property
```
Comment (GuilleEneas) [inline packages/jasmine-auto-spies/src/tests/observable-spy-utils.spec.ts:599]: I duplicated the tests for the getter, I didn't know what else to do to ensure that a getter will behave as a plain observable property

**Review by GuilleEneas — COMMENTED** [review]
I just left a couple of comments on topics I am not 100% secure

File: packages/auto-spies-core/src/create-auto-spy-from-class.ts (hunk 1 above)
Comment (shairez) [reply packages/auto-spies-core/src/create-auto-spy-from-class.ts:97]: I believe most of the testers will use either latest chrome or JSDOM.. so... I'm not sure if that should be a concern.

But.. you can always use `keys` if it's not too of a hassle

File: packages/jasmine-auto-spies/src/tests/observable-spy-utils.spec.ts (hunk 2 above)
Comment (shairez) [reply packages/jasmine-auto-spies/src/tests/observable-spy-utils.spec.ts:599]: I don't think there's a need to duplicate all of the tests.

Just one test that proves that this bug is fixed should be suffice. 

Can you please remove the rest of the tests and give the `describe` a more specific name to this issue? 

Thanks!

**Comment by GuilleEneas** [comment]
@shairez I change the tests and modified the loop. please take a look again 🤓

**Comment by shairez** [comment]
Amazing job @GuilleEneas !
Thanks a lot!

**Pull request merged**